src/transformers/models/levit/modeling_levit.py

```diff
@@ -0,0 +1,729 @@
+        self.activation = nn.Hardswish()
+        self.linear_down = MLPLayerWithBN(hidden_dim, input_dim)
+
+    def forward(self, hidden_state):
+        hidden_state = self.linear_up(hidden_state)
+        hidden_state = self.activation(hidden_state)
+        hidden_state = self.linear_down(hidden_state)
+        return hidden_state
+
+
+class LevitResidualBlock(nn.Module):
+    """
+    Residual Block for LeViT
+    """
+
+    def __init__(self, module, drop_rate):
+        super().__init__()
+        self.module = module
+        self.drop_rate = drop_rate
+
+    def forward(self, x):
+        if self.training and self.drop_rate > 0:
+            x = (
+                x
+                + self.module(x)
+                * torch.rand(x.size(0), 1, 1, device=x.device).ge_(self.drop_rate).div(1 - self.drop_rate).detach()
+            )
+            return x
+        else:
+            x = x + self.module(x)
+            return x
+
+
+class LevitEncoder(nn.Module):
+    """
+    LeViT Encoder Module
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.num_features = config.embed_dim[-1]
+        self.embed_dim = config.embed_dim
+
+        resolution = config.image_size // config.patch_size
+
+        self.patch_embeddings = LevitPatchEmbeddings(config)
+
+        self.blocks = []
+        config.down_ops.append([""])
+
+        for idx, (ed, kd, dpt, nh, attn_r, mlp_r, do) in enumerate(
+            zip(
+                config.embed_dim,
+                config.key_dim,
+                config.depth,
+                config.num_heads,
+                config.attention_ratio,
+                config.mlp_ratio,
+                config.down_ops,
+            )
+        ):
+            for _ in range(dpt):
+                self.blocks.append(
+                    LevitResidualBlock(LevitAttention(ed, kd, nh, attn_r, resolution), config.drop_path)
+                )
+                if mlp_r > 0:
+                    hidden_dim = ed * mlp_r
+                    self.blocks.append(LevitResidualBlock(LevitMLPBlock(ed, hidden_dim), config.drop_path))
+
+            if do[0] == "Subsample":
+                resolution_ = (resolution - 1) // do[5] + 1
+                self.blocks.append(
+                    LevitAttentionSubsample(
+                        *config.embed_dim[idx : idx + 2],
+                        key_dim=do[1],
+                        num_heads=do[2],
+                        attention_ratio=do[3],
+                        stride=do[5],
+                        resolution=resolution,
+                        resolution_=resolution_,
+                    )
+                )
+                resolution = resolution_
+                if do[4] > 0:
+                    hidden_dim = config.embed_dim[idx + 1] * do[4]
+                    self.blocks.append(
+                        LevitResidualBlock(LevitMLPBlock(config.embed_dim[idx + 1], hidden_dim), config.drop_path)
```

Comment:
I'm wondering whether this can be wrapped up in a LevitStage (see ConvNextStage as an example).

Typically, vision models consists of stages, each stage consisting of several layers.

Rather than modeling everything in the encoder, maybe it would become more readable to do it as in ConvNeXT.

Reply:
Done